aten/src/ATen/native/mps/operations/ScanKernel.mm

```diff
@@ -21,11 +21,174 @@
 #include <ATen/native/mps/ScanKernel_metallib.h>
 #endif
 
+// Utility function to get 2D grid dimensions for dispatch based on MLX implementation
+static std::pair<uint32_t, uint32_t> get_2d_grid_dims(const IntArrayRef& shape,
+                                                      const IntArrayRef& strides,
+                                                      size_t divisor) {
+  // Compute the 2d grid dimensions such that the total size of the grid is
+  // divided by divisor.
+  size_t grid_x = 1;
+  size_t grid_y = 1;
+
+  for (int i = 0; i < shape.size(); ++i) {
+    if (strides[i] == 0) {
+      continue; // Skip broadcasted dimensions
+    }
+
+    // No need to add this shape we can just remove it from the divisor.
+    if (divisor % shape[i] == 0) {
+      divisor /= shape[i];
+      continue;
+    }
+
+    if (grid_x * shape[i] < UINT32_MAX) {
+      grid_x *= shape[i];
+    } else {
+      grid_y *= shape[i];
+    }
+
+    if (divisor > 1) {
+      if (grid_x % divisor == 0) {
+        grid_x /= divisor;
+        divisor = 1;
+      } else if (grid_y % divisor == 0) {
+        grid_y /= divisor;
+        divisor = 1;
+      }
+    }
+  }
+
+  TORCH_CHECK(grid_y <= UINT32_MAX && grid_x <= UINT32_MAX && divisor <= 1,
+              "Unable to safely factor shape for grid dimensions.");
+
+  if (grid_y > grid_x) {
+    std::swap(grid_x, grid_y);
+  }
+
+  return std::make_pair(static_cast<uint32_t>(grid_x), static_cast<uint32_t>(grid_y));
+}
+
+static void scan_mps_impl(const Tensor& self, const Tensor& output, int64_t dim, const std::string& op_name) {
+  if (output.numel() == 0) {
+    return;
+  }
+
+  const int64_t ndim = self.dim();
+  const int64_t wrapped_dim = maybe_wrap_dim(dim, ndim);
+  const int64_t axis_size = self.size(wrapped_dim);
+
+  // Preprocess input tensor - ensure it's contiguous for Metal shaders
+  Tensor input_tensor = self;
+  bool input_needs_copy = !self.is_contiguous();
+
+  if (input_needs_copy) {
+    input_tensor = self.contiguous();
+  }
```

Comment:
`self.contiguous()` is a no-op if input is already contiguous, otherwise it returns contig tensor
```suggestion
  auto& input_tensor = self.contiguous();
```